Introduce Z3 types and utility functions for constraint generation

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80147
* #80110
* #80102
* #80095
* __->__ #80084

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

migeed-z committed:
Introduce Z3 types and utility functions for constraint generation

[ghstack-poisoned]

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"




[ghstack-poisoned]

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

[ghstack-poisoned]

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

[ghstack-poisoned]

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37357902).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37357902).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37357902).

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/util.py:
```diff
@@ -0,0 +1,27 @@
+from torch.fx.experimental.migrate_gradual_types.operation import op_leq  # type: ignore[import]
+
+
+def gen_tvar(curr):
```
Maybe add docstrings to the functions in this file?

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/z3_types.py:
```diff
@@ -0,0 +1,28 @@
+try:
```
This looks like a test. Can you move this to a file in https://github.com/pytorch/pytorch/tree/master/test/fx

It might also be good to wrap this stuff up in a `TestCase` instance to help give better interfacing to test runners and diagnostic information. See `test_dce_pass.py` for an example: https://github.com/pytorch/pytorch/blob/f749f86feec312b0efc9237d540973fa8414cd96/test/fx/test_dce_pass.py#L10

jamesr66a reviewed (commented):
Seems fine, but some small comments

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37357902).

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37357902).

migeed-z commented:
@pytorchbot merge -g

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37357902).

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Differential Revision: [D37357902](https://our.internmc.facebook.com/intern/diff/D37357902)

Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z commented:
@pytorchbot merge -g

migeed-z committed:
Update on "Introduce Z3 types and utility functions for constraint generation"


Create Z3 types. In particular, dynamic dimensions, dynamic tensor type and tensor types up to size 4. Note that for Z3 decidability reasons, we are using uninterpreted functions for tensor types, which means we must explicitly define tensor constructors with a concrete size (for now, upto size 4).  We defer lifting this requirement to future work.

[ghstack-poisoned]

migeed-z commented:
@pytorchbot merge -g

migeed-z commented:
@pytorchbot merge -g
